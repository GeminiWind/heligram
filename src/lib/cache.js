import { promisify } from 'util';
import redis from 'redis';

const client = redis.createClient('redis://cache');

const get = promisify(client.get).bind(client);
const set = promisify(client.set).bind(client);
const hset = promisify(client.hset).bind(client);
const hget = promisify(client.hget).bind(client);
const del = promisify(client.del).bind(client);
const hdel = promisify(client.hdel).bind(client);

export default {
  get,
  set,
  hset,
  hget,
  del,
  hdel,
};
