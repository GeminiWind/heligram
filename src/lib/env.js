import { config } from 'dotenv';

export default function loadEnv() {
  const envFilePath = process.env.NODE_ENV ? `env.${process.env.NODE_ENV}` : '.env.development';

  config({ path: envFilePath });
}
