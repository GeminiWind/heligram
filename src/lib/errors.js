import JsonApiError from './jsonApiError';

class InternalError extends JsonApiError {
  constructor(message) {
    let internalError = {
      title: 'InternalError',
      code: 'InternalError',
      status: 500,
      detail: 'InternalError',
    };
    if (typeof message === 'string') {
      internalError = {
        ...internalError,
        detail: message,
      };
    }
    if (typeof message === 'object') {
      internalError = {
        ...internalError,
        ...message,
      };
    }
    super(internalError);
  }
}

class NotFoundError extends JsonApiError {
  constructor(message) {
    let notFoundError = {
      title: 'NotFoundError',
      code: 'NotFoundError',
      status: 404,
      detail: 'NotFoundError',
    };
    if (typeof message === 'string') {
      notFoundError = {
        ...notFoundError,
        detail: message,
      };
    }
    if (typeof message === 'object') {
      notFoundError = {
        ...notFoundError,
        ...message,
      };
    }
    super(notFoundError);
  }
}

class BadRequestError extends JsonApiError {
  constructor(message) {
    let badRequestError = {
      title: 'BadRequestError',
      code: 'BadRequestError',
      status: 400,
      detail: 'BadRequestError',
    };
    if (typeof message === 'string') {
      badRequestError = {
        ...badRequestError,
        detail: message,
      };
    }
    if (typeof message === 'object') {
      badRequestError = {
        ...badRequestError,
        ...message,
      };
    }
    super(badRequestError);
  }
}

class MalformedError extends JsonApiError {
  constructor() {
    const malformedError = {
      title: 'MalformedError',
      code: 'MalformedError',
      status: 400,
      detail: 'Request body is malformed.',
    };
    super(malformedError);
  }
}

export {
  InternalError,
  NotFoundError,
  BadRequestError,
  MalformedError,
};
