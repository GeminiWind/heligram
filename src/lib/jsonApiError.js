class JsonApiError extends Error {
  constructor(error) {
    const errorMessage = `${error.title} ${error.detail}`;
    super(errorMessage);

    this.title = error.title;
    this.detail = error.detail;
    this.status = error.status;
    this.code = error.code;

    if (error.source) {
      this.source = error.source;
    }
    if (error.meta) {
      this.meta = error.meta;
    }

    // Ensure the name of this error is the same as the class name
    this.name = this.constructor.name;
    // This clips the constructor invocation from the stack trace.
    // It's not absolutely essential, but it does make the stack trace a little nicer.
    //  @see Node.js reference (bottom)
    Error.captureStackTrace(this, this.constructor);
  }

  toJSON() {
    return {
      status: this.status,
      code: this.code,
      title: this.title,
      detail: this.detail,
      ...(this.source ? { source: this.source } : {}),
      ...(this.meta ? { source: this.meta } : {}),
    };
  }
}

export default JsonApiError;
