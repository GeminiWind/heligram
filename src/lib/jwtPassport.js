import { Strategy as JwtStrategy, ExtractJwt } from 'passport-jwt';
import storageLibrary from './storageLibrary';

export default function jwtPassport(passport) {
  const options = {
    jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
    secretOrKey: process.env.APP_KEY,
  };

  passport.use(new JwtStrategy(options, (jwtPayload, done) => {
    const {
      email,
      scopes,
    } = jwtPayload;

    storageLibrary.findOne({
      Path: `user/${email}`,
    }).then((record) => {
      if (record) {
        const {
          Content: {
            password,
            ...userAttributesWithoutPassword
          },
        } = record;
        done(null, {
          ...userAttributesWithoutPassword,
          scopes,
        });
      } else done(null, false);
    }).catch(error => done(error, false));
  }));
}
