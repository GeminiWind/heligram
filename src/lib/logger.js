import moment from 'moment';
import path from 'path';
import { format, transports, createLogger } from 'winston';

const logger = createLogger({
  format: format.combine(
    format.splat(),
    format.timestamp({
      format: 'YYYY-MM-DD HH:mm:ss',
    }),
    format.printf(info => `${info.timestamp} ${info.level}: ${info.message}`),
  ),
  transports: [
    new transports.Console({
      format: format.combine(
        format.colorize(),
        format.timestamp({
          format: 'YYYY-MM-DD HH:mm:ss',
        }),
        format.printf(info => `${info.timestamp} ${info.level}: ${info.message}`),
      ),
    }),
    new transports.File({
      filename: path.resolve(__dirname, '../..', 'log', `${moment().format('YYYY-MM-DD')}.log`),
    }),
  ],
});

export default logger;
