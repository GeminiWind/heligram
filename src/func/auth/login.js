import jwt from 'jsonwebtoken';
import moment from 'moment';
import { isMatchingWithHashedPassword } from './helper';
import { NotFoundError, BadRequestError } from '../../lib/errors';

export default req => Promise.resolve(req)
  .then(validateRequest)
  .then(getUserByEmail)
  .then(verifyPassword)
  .then(generateToken)
  .then(returnResponse);

export function validateRequest(req) {
  const { schemaValidator, instrumentation } = req;
  const isValid = schemaValidator.validate('https://heligram.com/login-user+v1.json', req.body.data);

  if (!isValid) {
    instrumentation.error('Error in validate request %s', JSON.stringify(schemaValidator.errors, null, 2));

    throw new BadRequestError({
      source: schemaValidator.errors,
    });
  }

  return req;
}

export async function getUserByEmail(req) {
  const {
    instrumentation,
    storageLibrary,
    body: {
      data: {
        email,
      },
    },
  } = req;

  const user = await storageLibrary.findOne({
    Path: `user/${email}`,
  }).lean();

  if (!user) {
    instrumentation.error(`User with ${email} was not found`);

    throw new NotFoundError(`User with ${email} was not found`);
  }

  return {
    ...req,
    user,
  };
}

export function verifyPassword(req) {
  const {
    instrumentation,
    user: {
      Content: {
        password,
      },
    },
  } = req;

  if (!isMatchingWithHashedPassword(req.body.data.password, password)) {
    instrumentation.error('Password does not match');

    throw new BadRequestError('Password does not match.');
  }

  return req;
}

export function generateToken(req) {
  const expiresIn = parseInt('1800', 10); // ~30mins
  const token = jwt.sign(req.user, process.env.APP_KEY, {
    expiresIn, // in seconds
  });
  const expireAt = moment().add(30, 'm').unix();

  return {
    ...req,
    token: `Bearer ${token}`,
    expireAt,
  };
}

export function returnResponse(req) {
  return {
    statusCode: 200,
    body: {
      data: {
        type: 'session',
        attributes: {
          email: req.body.data.email,
          token: req.token,
          expireAt: req.expireAt,
        },
      },
    },
  };
}
