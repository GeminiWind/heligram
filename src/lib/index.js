import logger from './logger';
import storageLibrary from './storageLibrary';

export { logger, storageLibrary };
