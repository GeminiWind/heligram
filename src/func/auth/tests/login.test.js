import login, {
  validateRequest,
  getUserByEmail,
  verifyPassword,
  verifyRefreshToken,
  exchangeRefreshTokenToAccessToken,
  generateAccessToken,
  returnResponse,
} from '../login';
import { schemaValidator } from '../../../lib';
import { BadRequestError, NotFoundError } from '../../../lib/errors';

describe('Login Controller', () => {
  beforeAll(() => {
    process.env.APP_KEY = 'mock-app-key';
  });

  afterAll(() => {
    delete process.env.APP_KEY;
  });

  it('can validate request in case that request is matched with \'password\' grantType schema', () => {
    const request = {
      body: {
        data: {
          type: 'tokens',
          attributes: {
            grantType: 'password',
            email: 'validEmail@example.com',
            password: 'validPassword',
            scopes: 'user:profile create:chat read:chat',
          },
        },
      },
      instrumentation: {
        error: jest.fn(),
      },
      schemaValidator,
    };

    let error;

    try {
      validateRequest(request);
    } catch (e) {
      error = e;
    }

    expect(error).toBeFalsy();
  });

  it('can validate request in case that request is matched with \'refreshToken\' grantType schema', () => {
    const request = {
      body: {
        data: {
          type: 'tokens',
          attributes: {
            grantType: 'refreshToken',
            refreshToken: 'mock-refreshToken',
          },
        },
      },
      instrumentation: {
        error: jest.fn(),
      },
      schemaValidator,
    };

    let error;

    try {
      validateRequest(request);
    } catch (e) {
      error = e;
    }

    expect(error).toBeFalsy();
  });

  it('throw \'BadRequestError\'  in case that request does not matches with schema', () => {
    const request = {
      body: {
        data: {
          type: 'tokens',
          attributes: {
            email: 'invalidEmail',
            password: 'validPassword',
          },
        },
      },
      instrumentation: {
        error: jest.fn(),
      },
      schemaValidator,
    };

    let error;

    try {
      validateRequest(request);
    } catch (e) {
      error = e;
    }

    expect(error).toBeInstanceOf(BadRequestError);
    expect(error).toMatchSnapshot();
  });

  it('can get user by their email', async () => {
    const request = {
      body: {
        data: {
          type: 'tokens',
          attributes: {
            grantType: 'password',
            email: 'validEmail@example.com',
            password: 'validPassword',
            scopes: 'user:profile create:chat read:chat',
          },
        },
      },
      instrumentation: {
        error: jest.fn(),
      },
      schemaValidator,
      storageLibrary: {
        findOne: jest.fn().mockImplementation(() => ({
          cache: jest.fn().mockImplementation(() => ({
            lean: jest.fn().mockImplementation(() => Promise.resolve({
              Path: 'user/validEmail@example.com',
              Content: {
                email: 'validEmail@example.com',
                password: 'validPassword',
              },
            })),
          })),
        })),
      },
    };

    const state = await getUserByEmail(request);

    expect(state.user).toBeDefined();
    expect(state.user).toMatchSnapshot();
  });

  it('throw \'NotFoundError\' if user does not exist', async () => {
    const request = {
      body: {
        data: {
          type: 'tokens',
          attributes: {
            grantType: 'password',
            email: 'validEmail@example.com',
            password: 'validPassword',
            scopes: 'user:profile create:chat read:chat',
          },
        },
      },
      instrumentation: {
        error: jest.fn(),
      },
      schemaValidator,
      storageLibrary: {
        findOne: jest.fn().mockImplementation(() => ({
          cache: jest.fn().mockImplementation(() => ({
            lean: jest.fn().mockImplementation(() => Promise.resolve(false)),
          })),
        })),
      },
    };

    let error;

    try {
      await getUserByEmail(request);
    } catch (e) {
      error = e;
    }

    expect(error).toBeInstanceOf(NotFoundError);
    expect(error).toMatchSnapshot();
  });

  // FIXME: get real hashed password
  it('can verify password if password is correct', () => {
    const request = {
      body: {
        data: {
          type: 'tokens',
          attributes: {
            grantType: 'password',
            email: 'validEmail@example.com',
            password: 'temando@123',
          },
        },
      },
      user: {
        email: 'validEmail@example.com',
        password: '$2b$10$lmtfpvxcPqC/tjJdqe7T6e3s.xWA2KxUQLZsA794xZTk.cJSQeIxi',
      },
      instrumentation: {
        error: jest.fn(),
      },
    };

    const state = verifyPassword(request);

    expect(state).toBeDefined();
    expect(state).toMatchSnapshot();
  });

  it('can generate token for specified user', async () => {
    const request = {
      body: {
        data: {
          type: 'tokens',
          attributes: {
            grantType: 'password',
            email: 'validEmail@example.com',
            password: 'validPassword',
            scopes: 'user:profile create:chat read:chat',
          },
        },
      },
      user: {
        email: 'validEmail@example.com',
        password: 'notCorrect',
        scopes: 'user:profile create:chat read:chat',
      },
      cache: {
        set: jest.fn().mockImplementation(value => Promise.resolve(value)),
      },
    };

    const state = await generateAccessToken(request);

    expect(state).toHaveProperty('refreshToken');
    expect(state).toHaveProperty('accessToken');
    expect(state).toHaveProperty('expireAt');
  });

  it('throw \'BadRequestError\' if requested scope for user is not valid', async () => {
    const request = {
      body: {
        data: {
          type: 'tokens',
          attributes: {
            grantType: 'password',
            email: 'validEmail@example.com',
            password: 'validPassword',
            scopes: 'user:profile create:chat read:chat delete:user',
          },
        },
      },
      user: {
        email: 'validEmail@example.com',
        password: 'notCorrect',
        scopes: 'user:profile create:chat read:chat',
      },
      cache: {
        set: jest.fn().mockImplementation(value => Promise.resolve(value)),
      },
    };

    let error;

    try {
      await generateAccessToken(request);
    } catch (e) {
      error = e;
    }

    expect(error).toBeDefined();
    expect(error).toBeInstanceOf(BadRequestError);
  });

  it('can return response correctly', () => {
    const request = {
      accessToken: 'Berear mockToken',
      refreshToken: 'mock-refreshToken',
      expireAt: '123456',
    };

    const response = returnResponse(request);

    expect(response).toHaveProperty('statusCode', 200);
    expect(response).toHaveProperty('body');
    expect(response).toMatchSnapshot();
  });

  it('return original state if provided refreshToken is valid', async () => {
    const request = {
      cache: {
        get: jest.fn().mockImplementation(() => Promise.resolve(JSON.stringify({
          email: 'validEmail@example.com',
          scopes: 'user:profile create:chat read:chat',
        }))),
      },
      body: {
        data: {
          type: 'tokens',
          attributes: {
            grantType: 'refreshToken',
            refreshToken: 'mock-refreshToken',
          },
        },
      },
    };

    let error;

    try {
      await verifyRefreshToken(request);
    } catch (e) {
      error = e;
    }

    expect(error).toBeUndefined();
  });

  it('throw \'BadRequestError\' if provided refreshToken is invalid', async () => {
    const request = {
      cache: {
        get: jest.fn().mockImplementation(() => Promise.resolve(undefined)),
      },
      body: {
        data: {
          type: 'tokens',
          attributes: {
            grantType: 'refreshToken',
            refreshToken: 'invalid-refreshToken',
          },
        },
      },
    };

    let error;

    try {
      await verifyRefreshToken(request);
    } catch (e) {
      error = e;
    }

    expect(error).toBeInstanceOf(BadRequestError);
    expect(error).toMatchSnapshot();
  });

  it('success for the whole flow in case grantType is \'password\'', async () => {
    const request = {
      body: {
        data: {
          type: 'tokens',
          attributes: {
            grantType: 'password',
            email: 'validEmail@example.com',
            password: 'temando@123',
            scopes: 'user:profile create:chat read:chat',
          },
        },
      },
      instrumentation: {
        error: jest.fn(),
      },
      schemaValidator,
      storageLibrary: {
        findOne: jest.fn().mockImplementation(() => ({
          cache: jest.fn().mockImplementation(() => ({
            lean: jest.fn().mockImplementation(() => Promise.resolve({
              Path: 'user/validEmail@example.com',
              Content: {
                email: 'validEmail@example.com',
                password: '$2b$10$lmtfpvxcPqC/tjJdqe7T6e3s.xWA2KxUQLZsA794xZTk.cJSQeIxi',
                scopes: 'user:profile create:chat read:chat',
              },
            })),
          })),
        })),
      },
      cache: {
        set: jest.fn().mockImplementation(value => Promise.resolve(value)),
      },
    };

    const response = await login(request);

    expect(response).toHaveProperty('statusCode', 200);
    expect(response).toHaveProperty(['body', 'data', 'attributes', 'accessToken']);
    expect(response).toHaveProperty(['body', 'data', 'attributes', 'expireAt']);
    expect(response).toHaveProperty(['body', 'data', 'attributes', 'refreshToken']);
  });

  it('success for the whole flow in case grantType is \'refreshToken\'', async () => {
    const request = {
      body: {
        data: {
          type: 'tokens',
          attributes: {
            grantType: 'refreshToken',
            refreshToken: 'valid-refreshToken',
          },
        },
      },
      instrumentation: {
        error: jest.fn(),
      },
      schemaValidator,
      storageLibrary: {
        findOne: jest.fn().mockImplementation(() => ({
          cache: jest.fn().mockImplementation(() => ({
            lean: jest.fn().mockImplementation(() => Promise.resolve({
              Path: 'user/validEmail@example.com',
              Content: {
                email: 'validEmail@example.com',
                password: '$2b$10$lmtfpvxcPqC/tjJdqe7T6e3s.xWA2KxUQLZsA794xZTk.cJSQeIxi',
                scopes: 'user:profile create:chat read:chat',
              },
            })),
          })),
        })),
      },
      cache: {
        get: jest.fn().mockImplementation(() => Promise.resolve(JSON.stringify({
          email: 'validEmail@example.com',
          scopes: 'user:profile create:chat read:chat',
        }))),
      },
    };

    const response = await login(request);

    expect(response).toHaveProperty('statusCode', 200);
    expect(response).toHaveProperty(['body', 'data', 'attributes', 'accessToken']);
    expect(response).toHaveProperty(['body', 'data', 'attributes', 'expireAt']);
    expect(response).toHaveProperty(['body', 'data', 'attributes', 'refreshToken']);
  });
});
