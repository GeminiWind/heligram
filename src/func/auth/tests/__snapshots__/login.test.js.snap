// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Controller can get user by their email 1`] = `
Object {
  "Content": Object {
    "email": "validEmail@example.com",
    "password": "validPassword",
  },
  "Path": "user/validEmail@example.com",
}
`;

exports[`Login Controller can return response correctly 1`] = `
Object {
  "body": Object {
    "data": Object {
      "attributes": Object {
        "email": "validEmail@example.com",
        "expireAt": "123456",
        "token": "Berear mockToken",
      },
      "type": "session",
    },
  },
  "statusCode": 200,
}
`;

exports[`Login Controller can verify password if password is not correct 1`] = `[BadRequestError: BadRequestError Password does not match.]`;

exports[`Login Controller throw 'BadRequestError'  in case that request does not matches with schema 1`] = `[BadRequestError: BadRequestError BadRequestError]`;

exports[`Login Controller throw 'NotFoundError' if user does not exist 1`] = `[NotFoundError: NotFoundError User with validEmail@example.com was not found]`;
